File: server/src/routes/fleet.js
const router = require('express').Router();

// In-memory simulator state
const SCENES = {
  normal_operation:          { rpm:1480, hydraulic:218, fuel:72, temp:82, tilt:2.1, seatbelt:true,  workers:0, idle:8,  speed:4.2 },
  excessive_idle:            { rpm:820,  hydraulic:170, fuel:68, temp:75, tilt:0.5, seatbelt:true,  workers:0, idle:58, speed:0   },
  worker_proximity_breach:   { rpm:1400, hydraulic:210, fuel:65, temp:80, tilt:2.0, seatbelt:true,  workers:2, idle:4,  speed:3.0 },
  seatbelt_violation_moving: { rpm:1600, hydraulic:235, fuel:70, temp:83, tilt:3.0, seatbelt:false, workers:0, idle:2,  speed:6.5 },
  slope_instability:         { rpm:1350, hydraulic:205, fuel:60, temp:78, tilt:18.5,seatbelt:true,  workers:0, idle:3,  speed:1.5 },
  operator_fatigue:          { rpm:1200, hydraulic:190, fuel:55, temp:77, tilt:2.5, seatbelt:true,  workers:0, idle:20, speed:2.0 },
  cold_start_abuse:          { rpm:2210, hydraulic:288, fuel:98, temp:45, tilt:1.0, seatbelt:true,  workers:0, idle:1,  speed:1.0 },
  geo_fence_breach:          { rpm:1500, hydraulic:215, fuel:66, temp:81, tilt:2.0, seatbelt:true,  workers:0, idle:5,  speed:8.0 },
  night_low_visibility:      { rpm:1300, hydraulic:200, fuel:58, temp:72, tilt:2.2, seatbelt:true,  workers:1, idle:10, speed:2.5 },
};

const GPS = {
  EXC001: { lat:17.4501, lon:78.3821 }, EXC002: { lat:17.4515, lon:78.3835 },
  LDR001: { lat:17.4490, lon:78.3810 }, LDR002: { lat:17.4525, lon:78.3800 },
};
const TASKS = { EXC001:'Trenching - Sector 4', EXC002:'Grading - Sector 2', LDR001:'Loading - Pit A', LDR002:'Idle' };
const MACHINES = ['EXC001','EXC002','LDR001','LDR002'];
const machineScenes = { EXC001:'normal_operation', EXC002:'excessive_idle', LDR001:'worker_proximity_breach', LDR002:'normal_operation' };

function machineStatus(mid) {
  const t = SCENES[machineScenes[mid]] || SCENES.normal_operation;
  const alerts = [];
  if (!t.seatbelt && t.speed>1) alerts.push('SEATBELT_VIOLATION');
  if (t.workers>0)              alerts.push('PROXIMITY_BREACH');
  if (t.tilt>15)                alerts.push('SLOPE_INSTABILITY');
  if (t.idle>30)                alerts.push('EXCESSIVE_IDLE');
  const health = alerts.some(a=>['SLOPE_INSTABILITY','SEATBELT_VIOLATION'].includes(a)) ? 'critical' : alerts.length ? 'warning' : 'good';
  return { machine_id:mid, ...GPS[mid], status: t.idle>30?'idle':'active',
    task: TASKS[mid], rpm:t.rpm, fuel_level:t.fuel, health, active_alerts:alerts,
    telemetry:{ seatbelt:t.seatbelt, workers_in_zone:t.workers, idling_time:t.idle, tilt_angle:t.tilt, hydraulic_pressure:t.hydraulic } };
}

// GET /fleet/map
router.get('/map', (req, res) => {
  const machines = MACHINES.map(machineStatus);
  res.json({ site_id: req.query.site_id||'site_1', machines, total:machines.length,
    active: machines.filter(m=>m.status==='active').length,
    alerts: machines.reduce((s,m)=>s+m.active_alerts.length,0) });
});

// GET /fleet/machines
router.get('/machines', (req, res) => {
  res.json(Object.fromEntries(MACHINES.map(m=>[m, machineStatus(m)])));
});

// POST /fleet/scene
router.post('/scene', (req, res) => {
  const { machine_id='EXC001', scene_name } = req.query;
  if (!SCENES[scene_name]) return res.status(400).json({ error:'Unknown scene', available: Object.keys(SCENES) });
  machineScenes[machine_id] = scene_name;
  res.json({ machine_id, scene: scene_name, status:'loaded', telemetry: SCENES[scene_name] });
});

// POST /fleet/fault
router.post('/fault', (req, res) => {
  const { machine_id='EXC001', key, value } = req.query;
  if (machineScenes[machine_id] && SCENES[machineScenes[machine_id]]) {
    SCENES[machineScenes[machine_id]][key] = isNaN(value) ? value==='true' : Number(value);
  }
  res.json({ machine_id, fault: key, value, injected:true });
});

module.exports = router;

File: api/fleet.py
"""
Fleet API — live machine positions, health summary, supervisor view.
Uses ExcavatorSimulator for demo; replace with real telemetry adapter in production.
"""
from __future__ import annotations

import sys
from pathlib import Path
from fastapi import APIRouter

ROOT = Path(__file__).resolve().parent.parent
if str(ROOT) not in sys.path:
    sys.path.insert(0, str(ROOT))

from sim.excavator import ExcavatorSimulator, EXCAVATOR_SCENES

router = APIRouter()

# One simulator per machine for demo
_sims: dict[str, ExcavatorSimulator] = {
    "EXC001": ExcavatorSimulator(),
    "EXC002": ExcavatorSimulator(),
    "LDR001": ExcavatorSimulator(),
    "LDR002": ExcavatorSimulator(),
}
_sims["EXC001"].load_scene("normal_operation")
_sims["EXC002"].load_scene("excessive_idle")
_sims["LDR001"].load_scene("worker_proximity_breach")
_sims["LDR002"].load_scene("normal_operation")

# Static GPS offsets per machine (in real system these come from GPS hardware)
_GPS: dict[str, tuple[float, float]] = {
    "EXC001": (17.4501, 78.3821),
    "EXC002": (17.4515, 78.3835),
    "LDR001": (17.4490, 78.3810),
    "LDR002": (17.4525, 78.3800),
}

_TASKS: dict[str, str] = {
    "EXC001": "Trenching - Sector 4",
    "EXC002": "Grading - Sector 2",
    "LDR001": "Loading - Pit A",
    "LDR002": "Idle",
}


def _machine_status(machine_id: str) -> dict:
    sim = _sims.get(machine_id)
    if not sim:
        return {}
    t = sim.get_telemetry()
    lat, lon = _GPS.get(machine_id, (17.45, 78.38))

    idle_min = t.get("idling_time", 0)
    seatbelt = t.get("seatbelt", True)
    workers = t.get("workers_in_zone", 0)
    tilt = t.get("tilt_angle", 2.0)

    alerts = []
    if not seatbelt:
        alerts.append("SEATBELT_VIOLATION")
    if idle_min > 30:
        alerts.append("EXCESSIVE_IDLE")
    if workers > 0:
        alerts.append("PROXIMITY_BREACH")
    if tilt > 15:
        alerts.append("SLOPE_INSTABILITY")

    health = "critical" if "SLOPE_INSTABILITY" in alerts or "SEATBELT_VIOLATION" in alerts \
        else "warning" if alerts else "good"

    return {
        "machine_id": machine_id,
        "lat": lat,
        "lon": lon,
        "status": "idle" if idle_min > 30 else "active",
        "task": _TASKS.get(machine_id, "Unknown"),
        "rpm": t.get("rpm", 0),
        "fuel_level": t.get("fuel_level", 72),
        "health": health,
        "active_alerts": alerts,
        "telemetry": {
            "seatbelt": seatbelt,
            "workers_in_zone": workers,
            "idling_time": idle_min,
            "tilt_angle": tilt,
            "hydraulic_pressure": t.get("hydraulic_pressure", 220),
        },
    }


@router.get("/map")
def get_map(site_id: str = "site_1"):
    machines = [_machine_status(mid) for mid in _sims]
    return {
        "site_id": site_id,
        "machines": machines,
        "total": len(machines),
        "active": sum(1 for m in machines if m.get("status") == "active"),
        "alerts": sum(len(m.get("active_alerts", [])) for m in machines),
    }


@router.get("/machines")
def get_machines(site_id: str = "site_1"):
    return {mid: _machine_status(mid) for mid in _sims}


@router.post("/scene")
def set_scene(machine_id: str, scene_name: str):
    """Load a simulator scene for demo purposes."""
    sim = _sims.get(machine_id)
    if not sim:
        return {"error": f"Unknown machine {machine_id}"}
    try:
        sim.load_scene(scene_name)
        return {"machine_id": machine_id, "scene": scene_name, "status": "loaded",
                "telemetry": sim.get_telemetry()}
    except ValueError as e:
        return {"error": str(e), "available": list(EXCAVATOR_SCENES)}


@router.post("/fault")
def inject_fault(machine_id: str, key: str, value: str):
    """Inject a fault into a machine for demo."""
    sim = _sims.get(machine_id)
    if not sim:
        return {"error": f"Unknown machine {machine_id}"}
    # coerce value
    v: bool | float | int | str = value
    if value.lower() == "true":
        v = True
    elif value.lower() == "false":
        v = False
    else:
        try:
            v = float(value)
        except ValueError:
            pass
    sim.inject_fault(key, v)
    return {"machine_id": machine_id, "fault": key, "value": v, "injected": True}
